src/menu/index.tsx

```diff
@@ -0,0 +1,275 @@
+
+		const itemHeight = icache.getOrSet('itemHeight', offscreenHeight(offscreenMenuItem));
+
+		itemHeight && icache.set('menuHeight', itemsInView * itemHeight);
+	}
+
+	const selectedValue = icache.get('value');
+	const computedActiveIndex =
+		activeIndex === undefined ? icache.getOrSet('activeIndex', 0) : activeIndex;
+
+	function setActiveIndex(index: number) {
+		if (onActiveIndexChange) {
+			onActiveIndexChange(index);
+		} else {
+			icache.set('activeIndex', index);
+		}
+	}
+
+	function setValue(value: string) {
+		icache.set('value', value);
+		onValue(value);
+	}
+
+	function onKeyDown(event: KeyboardEvent) {
+		event.stopPropagation();
+
+		switch (event.which) {
+			case Keys.Enter:
+			case Keys.Space:
+				event.preventDefault();
+				const activeItem = options[computedActiveIndex];
+				!activeItem.disabled && setValue(activeItem.value);
+				break;
+			case Keys.Down:
+				event.preventDefault();
+				setActiveIndex((computedActiveIndex + 1) % options.length);
+				break;
+			case Keys.Up:
+				event.preventDefault();
+				setActiveIndex((computedActiveIndex - 1 + options.length) % options.length);
+				break;
+			case Keys.Escape:
+				event.preventDefault();
+				onRequestClose && onRequestClose();
+				break;
+			default:
+				const newIndex = getComputedIndexFromInput(event.key);
+				if (newIndex !== undefined) {
+					setActiveIndex(newIndex);
+				}
+		}
+	}
+
+	function getComputedIndexFromInput(key: string) {
+		const existingTimer = icache.get('resetInputTextTimer');
+		let inputText = icache.getOrSet('inputText', '') + `${key}`;
+		existingTimer && clearTimeout(existingTimer);
+
+		const resetTextTimeout = setTimeout(() => {
+			icache.set('inputText', '');
+		}, 800);
+
+		icache.set('resetInputTextTimer', resetTextTimeout);
+		icache.set('inputText', inputText);
+
+		return findIndex(
+			options,
+			({ disabled, value, label }) =>
+				!disabled && (label || value).toLowerCase().indexOf(inputText.toLowerCase()) === 0
+		);
+	}
+
+	function onActive(index: number, itemDimensions: DimensionResults) {
+		const { position: itemPosition, size: itemSize } = itemDimensions;
+		const { position: rootPosition, size: rootSize } = dimensions.get('root');
+		if (itemPosition.bottom > rootPosition.bottom) {
+			const numInView = Math.ceil(rootSize.height / itemSize.height);
+			icache.set('itemToScroll', Math.max(index - numInView + 1, 0));
+		} else if (itemPosition.top < rootPosition.top) {
+			icache.set('itemToScroll', index);
+		}
+	}
+
+	const itemToScroll = icache.get('itemToScroll');
+	const menuHeight = icache.get('menuHeight');
+	const rootStyles = menuHeight ? { maxHeight: `${menuHeight}px` } : {};
+	const shouldFocus = focus.shouldFocus();
+	const classes = theme.classes(css);
+
+	return (
+		<div
+			key="root"
+			classes={classes.menu}
+			tabIndex={focusable ? 0 : -1}
+			onkeydown={onKeyDown}
+			focus={() => shouldFocus}
+			onfocus={onFocus}
+			onblur={onBlur}
+			styles={rootStyles}
+			role="menu"
```

Comment:
Same comment as earlier: `role="menu"` is not correct for every use case for which we intend this widget, and it would be best to allow each implementation to specify its own accessibility-related properties.

Reply:
can we think of any other scenario's other than `menu` and `listbox` that the menu would ever be reasonably used for? I think we should really try and avoid basically the aria html forwarding bags we've done elsewhere if this is designed to be easily used by an end user.

Reply:
agreed, menu / listbox are the only cases i can think of